### PR TITLE
[Merged by Bors] - feat(NumberField/CanonicalEmbedding): put an euclidean space structure on the mixed space

### PR DESCRIPTION
**Pull request by xroblot**
This PR defines the natural euclidian space stucture on the mixed space `ℝ^r₁ × ℂ^r₂` of a number field of signature of `K` and the map between the `Euclidean.mixedSpace K` and `mixedSpace`. Proves many properties of this map. In particular, it is volume preserving. 

This PR is part of the proof of the Analytic Class Number Formula.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
Part of this is PR is also in #18234, right?

**Comment by xroblot** [comment]
> Part of this is PR is also in #18234, right?

Sorry. I forgot to finish cleaning up before removing WIP. Let me put it back for the time being. 

(I am getting a bit lost with all these PRs open at the same time 😅 )

**Comment by riccardobrasca** [comment]
> > Part of this is PR is also in #18234, right?
> 
> Sorry. I forgot to finish cleaning up before removing WIP. Let me put it back for the time being.
> 
> (I am getting a bit lost with all these PRs open at the same time 😅 )

OK, let's merge the other one, so the diff will become smaller (modulo conflicts).

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 1]
```diff
@@ -769,6 +770,89 @@ theorem fundamentalDomain_idealLattice :
...
+
+/-- The mixed space `ℝ^r₁ × ℂ^r₂`, with `(r₁, r₂)` the signature of `K`, as an Euclidean space. -/
+protected abbrev mixedSpace :=
+    (WithLp 2 ((EuclideanSpace ℝ {w : InfinitePlace K // IsReal w}) ×
+      (EuclideanSpace ℂ {w : InfinitePlace K // IsComplex w})))
+
+instance : Ring (euclidean.mixedSpace K) := by
+  have : Ring (EuclideanSpace ℝ {w : InfinitePlace K // IsReal w}) := Pi.ring
+  have : Ring (EuclideanSpace ℂ {w : InfinitePlace K // IsComplex w}) := Pi.ring
+  exact Prod.instRing
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:787]:
```suggestion
instance : Ring (euclidean.mixedSpace K) :=
  have : Ring (EuclideanSpace ℝ {w : InfinitePlace K // IsReal w}) := Pi.ring
  have : Ring (EuclideanSpace ℂ {w : InfinitePlace K // IsComplex w}) := Pi.ring
  inferInstanceAs (Ring (_ × _))
```
It is probably more robust.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 2]
```diff
@@ -769,6 +770,89 @@ theorem fundamentalDomain_idealLattice :
...
+
+instance : MeasurableSpace (euclidean.mixedSpace K) := borel _
+
+instance : BorelSpace (euclidean.mixedSpace K) := ⟨rfl⟩
+
+variable [NumberField K]
+
+open Classical in
+instance : T2Space (euclidean.mixedSpace K) := Prod.t2Space
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:796]: Here `inferInstance` works, so this is maybe useless.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 3]
```diff
@@ -769,6 +770,89 @@ theorem fundamentalDomain_idealLattice :
...
+/-- An orthonormal basis of the euclidean mixed space. -/
+def stdOrthonormalBasis : OrthonormalBasis (index K) ℝ (euclidean.mixedSpace K) :=
+  OrthonormalBasis.prod (EuclideanSpace.basisFun _ ℝ)
+    ((Pi.orthonormalBasis fun _ ↦ Complex.orthonormalBasisOneI).reindex (Equiv.sigmaEquivProd _ _))
+
+open Classical in
+theorem stdOrthonormalBasis_map_eq :
+    (euclidean.stdOrthonormalBasis K).toBasis.map (toMixed K).toLinearEquiv =
+      mixedEmbedding.stdBasis K := by
+  ext _ _ <;> rfl
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:819]:
```suggestion
  ext <;> rfl
```

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean (hunk 3 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean]: Thanks for the review!

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 4]
```diff
@@ -216,6 +217,25 @@ theorem _root_.NumberField.mixedEmbedding_injective [NumberField K] :
...
+
+open Classical in
+instance : IsAddHaarMeasure (volume : Measure (mixedSpace K)) :=
+  prod.instIsAddHaarMeasure volume volume
+
+open Classical in
+instance : NoAtoms (volume : Measure (mixedSpace K)) := by
+  obtain ⟨w⟩ := (inferInstance : Nonempty (InfinitePlace K))
+  by_cases hw : IsReal w
+  · exact @prod.instNoAtoms_fst _ _ _ _ volume volume _ (pi_noAtoms ⟨w, hw⟩)
```
[the remark this answers is not in the document]
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean]: Right, I need to clean this up

**Pull request closed without merging**